kong/tracing/request_id.lua

```diff
@@ -0,0 +1,108 @@
+local to_hex = require "resty.string".to_hex
+local base = require "resty.core.base"
+
+local get_request = base.get_request
+local var = ngx.var
+
+
+-- Request ID types
+-- higher value means higher priority
+-- CORR  = correlation ID
+-- TRACE = trace ID
+-- INIT  = the initial value
+local TYPES = {
+  INIT  = 1,
+  TRACE = 2,
+  CORR  = 3,
+}
+
+local NGX_VAR_PHASES = {
+  set           = true,
+  rewrite       = true,
+  access        = true,
+  content       = true,
+  header_filter = true,
+  body_filter   = true,
+  log           = true,
+  balancer      = true,
+}
+
+
+local function get_ctx_req_id()
+  return ngx.ctx.request_id and ngx.ctx.request_id.id
+end
+
+
+local function set_ctx_req_id(id, type)
+  ngx.ctx.request_id = {
+    id = id,
+    type = type,
+  }
+end
+
+
+local function get()
+  if not get_request() then
+    return nil, "no request found"
+  end
+
+  local rid = get_ctx_req_id()
+
+  if not rid then
+    -- first access to the request id for this request:
+    -- try to initialize with the value of $kong_request_id
+    local ok
+    ok, rid = pcall(function() return ngx.var.kong_request_id end)
+
+    if ok and rid then
+      set_ctx_req_id(rid, TYPES.INIT)
+    end
+  end
+
+  return rid
+end
+
+
+local function set(id, type)
+  -- phase and input checks
+  local phase = ngx.get_phase()
+  assert(NGX_VAR_PHASES[phase], "cannot set request_id in '" .. phase .. "' phase")
+  if not id or not type then
+    return nil, "both id and type are required"
+  end
+
+  -- priority check
+  local old_type = ngx.ctx.request_id and ngx.ctx.request_id.type or TYPES.INIT
+  if type < old_type then
+    return nil, "ignoring set for request_id of type: " .. type ..
+                ", less prioritary than current type: " .. ngx.ctx.request_id.type
+  end
```

Comment:
This seems to be the only _business logic_ of this function. Consider moving this to a separate function and unittest the precedence of `type` for all the known cases (INIT, TRACE, CORR). 

Also a nil check for `ngx.ctx.request_id.type` can be useful here